src/coffea/nanoevents/transforms.py

```diff
@@ -282,20 +281,494 @@ def index_range(stack):
+        return nested_local2global(array, begin.layout.offsets.data)
+    elif awkward.sum(array) == 0:
+        return awkward.flatten(
+            awkward.full_like(
+                awkward.values_astype(begin, "int64")[:, :, numpy.newaxis], "-1"
+            ),
+            axis=1,
+        )
+
+
+def nested_local2global(array, target_offsets_raw):
+    counts2 = awkward.flatten(awkward.num(array, axis=2), axis=1)
+    # flat_index = awkward.values_astype(awkward.flatten(awkward.local_index(array), axis=2),"int64")
+    flat_index = awkward.values_astype(awkward.flatten(array, axis=2), "int64")
+
+    target_offsets = awkward.values_astype(target_offsets_raw, "int64")
+
+    flat_index = flat_index.mask[flat_index >= 0] + target_offsets[:-1]
+    flat_index = flat_index.mask[flat_index < target_offsets[1:]]
+    out = ensure_array(awkward.flatten(awkward.fill_none(flat_index, -1), axis=None))
+    if out.dtype != numpy.int64:
+        raise RuntimeError
+
+    nested_global = awkward.unflatten(out, counts2, axis=0)
+    return nested_global
+
+
+def nested_local2global_stack(stack):
+    target_offsets_raw = stack.pop()
+    array = stack.pop()
+    counts1 = awkward.num(array, axis=1)
+    counts2 = awkward.flatten(awkward.num(array, axis=2), axis=1)
+
+    if awkward.sum(counts2) == 0:  # Empty indices
+        # print('Empty Indices')
+        nested_global = array
+    else:
+
+        # flat_index = awkward.values_astype(awkward.flatten(awkward.local_index(array), axis=2),"int64")
+        flat_index = awkward.values_astype(awkward.flatten(array, axis=2), "int64")
+
+        target_offsets = awkward.values_astype(target_offsets_raw, "int64")
+
+        flat_index = flat_index.mask[flat_index >= 0] + target_offsets[:-1]
+        flat_index = flat_index.mask[flat_index < target_offsets[1:]]
+        out = ensure_array(
+            awkward.flatten(awkward.fill_none(flat_index, -1), axis=None)
+        )
+
+        # if out.dtype != numpy.int64:
+        #     raise RuntimeError
+
+        nested_global_flat = awkward.unflatten(out, counts2, axis=0)
+        nested_global = awkward.unflatten(nested_global_flat, counts1, axis=0)
+    # print(nested_global)
+    stack.append(nested_global)
+
+
+def nested_local2global_form(array_form, target_offsets_form):
+    if not array_form["class"].startswith("ListOffset"):
+        raise RuntimeError
+    if not target_offsets_form["class"].startswith("NumpyArray"):
+        raise RuntimeError
+    form = copy.deepcopy(array_form)
+    form["content"]["content"]["primitive"] = "int64"
+    form["content"]["content"]["form_key"] = concat(
+        array_form["form_key"],
+        target_offsets_form["form_key"],
+        "!nested_local2global_stack",
+        "!content",
+        "!content",
+    )
+    return form
+
+
+def get_index_ranges(begin, end):
+    begin_end = awkward.concatenate(
+        (begin[:, :, numpy.newaxis], end[:, :, numpy.newaxis]), axis=2
+    )
+
+    @numba.njit
```

Comment:
Having this be a local function seems like a possible source of slowdown / confusion, please move this into the module namespace!